### PR TITLE
Introduce iterable stream json marshaler

### PR DESCRIPTION
**Pull request by ogaca-dd**
### What does this PR do?

This is the second part of reducing the memory used by flushSeries.
For general context see https://github.com/DataDog/datadog-agent/pull/10046.

This PR introduces `IterableSeries` which allow to flush series and serialized them at the same time.
Usage
```go
series := NewIterableSeries(func(*Serie){}, 100)
go func() {
  defer m.IterationStopped()
  series.WriteHeader(stream)
  for MoveNext() {
	WriteCurrentItem(stream)
  }
 series.WriteFooter(stream)
}()
series.Append(series1)
series.Append(series2)
series.Append(series3)
series.SenderStopped() // No more serie sent
```
`IterableSeries` aims to replace `Series`. As `IterableSeries` can be serialized while series are added to it, it does not need to contain all the series before the serialization unlike (`Series`).

Note: The code of this PR is not yet used.

`IterableStreamJSONMarshalerAdapter` allows to minimize the number of changes in the serialization logic (Event and service check code remain unchange). 

### Describe how to test/QA your changes

Nothing to QA in this card as code is not yet used.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Comment by djmitche** [comment]
Typically an iterator approach would use callbacks, rather than channels, to perform the iteration.  Is there any particular reason to mix iteration and channels here?

File: pkg/serializer/marshaler/marshaler.go [hunk 1]
```diff
@@ -52,6 +52,25 @@ type StreamJSONMarshaler interface {
 	DescribeItem(i int) string
 }
 
+// IterableStreamJSONMarshaler is an interface for iterable metrics that are able to
+// serialize themselves in a stream.
+// Expected usage:
+//
+//  defer m.IterationStopped()
+//	m.WriteHeader(stream)
+//	for MoveNext() {
```
Comment (ogaca-dd) [inline pkg/serializer/marshaler/marshaler.go:61]: `m.MoveNext`

**Comment by ogaca-dd** [comment]
@djmitche 

> Typically an iterator approach would use callbacks, rather than channels, to perform the iteration. Is there any particular reason to mix iteration and channels here?

Indeed callback is the typical approach with an iterator. There are 2 reasons why I have used a channel:
- Flushing series blocks the aggregator. Having a channel allow `aggregator.Flush` to return before all series are serialized in another goroutine (Same as today).
- `BuildWithOnErrItemTooBigPolicy` does not fit well with a callback and I have tried to minimize the changes.

File: pkg/metrics/iterable_series.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+func (series *IterableSeries) DescribeCurrentItem() string {
+	if series.current == nil {
+		return "nil serie"
+	}
+	return describeItem(series.current)
+}
+
+// MoveNext advances to the next element.
+// Returns false for the end of the iteration.
+func (series *IterableSeries) MoveNext() bool {
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply pkg/metrics/iterable_series.go:84]: I don't know if there is a standard way to name iterator in Go. I have chosen the .Net / Java one.

File: pkg/metrics/iterable_series.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+package metrics
+
+import (
+	"errors"
+
+	jsoniter "github.com/json-iterator/go"
+)
+
+// IterableSeries represents an iterable collection of Serie.
+// Serie can be appened to IterableSeries while IterableSeries is serialized
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply pkg/metrics/iterable_series.go]: Fixed

File: pkg/metrics/iterable_series.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+func (series *IterableSeries) WriteHeader(stream *jsoniter.Stream) error {
+	return writeHeader(stream)
+}
+
+// WriteFooter prints the payload footer for this type
+func (series *IterableSeries) WriteFooter(stream *jsoniter.Stream) error {
+	return writeFooter(stream)
+}
+
+// WriteCurrentItem prints the json representation of an item
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply pkg/metrics/iterable_series.go]: Note: I have copy pasted the comment from https://github.com/DataDog/datadog-agent/blob/5c3be2cba7105e4efacb01bdcd371fbfd4e500a2/pkg/metrics/series.go#L535.
Fixed

File: pkg/serializer/marshaler/marshaler_adapter.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+func (a *IterableStreamJSONMarshalerAdapter) WriteCurrentItem(j *jsoniter.Stream) error {
+	return a.marshaler.WriteItem(j, a.index)
+}
+
+// DescribeCurrentItem returns a text description
+func (a *IterableStreamJSONMarshalerAdapter) DescribeCurrentItem() string {
+	return a.marshaler.DescribeItem(a.index)
+}
+
+// MoveNext move the the next value. `HasNextValue` must return true.
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply pkg/serializer/marshaler/marshaler_adapter.go]: Fixed

File: pkg/serializer/serializer.go [hunk 6]
```diff
@@ -296,6 +302,22 @@ func (s *Serializer) SendServiceChecks(sc marshaler.StreamJSONMarshaler) error {
 	return s.Forwarder.SubmitServiceChecks(serviceCheckPayloads, extraHeaders)
 }
 
+// SendIterableSeries serializes a list of Serie and sends the payload to the forwarder
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply pkg/serializer/serializer.go]: I have updated to `SendIterableSeries serializes a list of series and sends the payload to the forwarder` to be consistent with other comments.

File: pkg/metrics/iterable_series.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+		current:             nil,
+	}
+}
+
+// Append appends a serie
+func (series *IterableSeries) Append(serie *Serie) {
+	series.callback(serie)
+	select {
+	case series.c <- serie:
+	case <-series.receiverStoppedChan:
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply pkg/metrics/iterable_series.go:42]: I have added a comment in the code.

File: pkg/metrics/iterable_series.go [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+	}
+}
+
+// SenderStopped must be called when sender stop calling Append.
+func (series *IterableSeries) SenderStopped() {
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply pkg/metrics/iterable_series.go:47]: `SenderStopped` is more an event to tell "I stop sending new value" whereas `StopSender()` seems to me more like the action of stopping like stopping a timer.

File: pkg/metrics/iterable_series.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+	close(series.c)
+}
+
+// IterationStopped must be called when the receiver stops calling MoveNext.
+// This function prevents the case when the receiver stops iterating
+// because of an error and so blocks the sender forever
+func (series *IterableSeries) IterationStopped() {
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply pkg/metrics/iterable_series.go:55]: Same comment as for `SenderStopped` but I am opened to other suggestions.

[the remark this answers is not in the document]
Comment (ogaca-dd) [reply pkg/metrics/iterable_series.go:55]: Note: I am trying removing `IterationStopped` and `SenderStopped` in a later PR.

**Review by remeh — APPROVED** [review]

**Pull request merged**